### PR TITLE
[Merged by Bors] - feat(LinearAlgebra/Matrix): define projective special linear group

### PR DESCRIPTION
**Pull request by shuxuezhuyi**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by alexjbest** [comment]
Do you have plans to prove theorems about `PSL` using this definition?

**Comment by shuxuezhuyi** [comment]
> Do you have plans to prove theorems about `PSL` using this definition?

Yes, but I will not prove many theorems about `PSL` since I only know a little about algebra. I only need `PSL(2, ℝ)` to define the Fuchs group.

**Comment by alexjbest** [comment]
Ok sounds good. Looking forward to seeing some Fuchsian groups. 

maintainer merge

**Comment by alexjbest** [comment]
It would also be good to add some API like we have for `SpecialLinearGroup`
so things like
```
Coe (Matrix.ProjectiveSpecialLinearGroup n R) (Matrix n n R)
```
lemmas `ext`, `ext_iff` etc

**Comment by alexjbest** [comment]
Actually thinking about it more, I'm not sure what lemmas even make sense, we don't have an actual coercion to matrices, so maybe this is fine like it is.

File: Mathlib/LinearAlgebra/Matrix/ProjectiveSpecialLinearGroup.lean [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+universe u v
+
+open Matrix LinearMap
+
+open scoped MatrixGroups
+
+variable (n : Type u) [DecidableEq n] [Fintype n] (R : Type v) [CommRing R]
+
+/-- A projective special linear group is the quotient of a special linear group by its center.-/
+def ProjectiveSpecialLinearGroup :=
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/ProjectiveSpecialLinearGroup.lean:23]: I think this should be an `abbrev`, then you can delete the instances below

**Comment by shuxuezhuyi** [comment]
I suspect that it might be problematic to instance a coercion from `PSL` to `SL`, since there is no canonical embedding from `PSL` to `SL`.
Actually, I defined a function `sl : PSL → SL`:
```lean
noncomputable def sl (a : ProjectiveSpecialLinearGroup n R) :
    SpecialLinearGroup n R := Classical.choose a.exists_rep
```
And I used it to instance `MulAction PSL(2, ℝ) ℍ`. But I am also trying to see if I can avoid using it.

File: Mathlib/LinearAlgebra/Matrix/ProjectiveSpecialLinearGroup.lean (hunk 1 above)
Comment (shuxuezhuyi) [reply Mathlib/LinearAlgebra/Matrix/ProjectiveSpecialLinearGroup.lean:23]: Amazing!

**Comment by eric-wieser** [comment]
That instance looks like a bad idea to me

File: Mathlib/LinearAlgebra/Matrix/ProjectiveSpecialLinearGroup.lean [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/-
+Copyright (c) 2023 Wen Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Wen Yang
+-/
+import Mathlib.LinearAlgebra.Matrix.SpecialLinearGroup
+
+/-!
+# Projective Special Linear Group
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/ProjectiveSpecialLinearGroup.lean:9]: This docstring should mention the notation and which `open scoped` is needed to obtain it.

File: Mathlib/LinearAlgebra/Matrix/ProjectiveSpecialLinearGroup.lean [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+universe u v
+
+open Matrix LinearMap
+
+open scoped MatrixGroups
+
+variable (n : Type u) [DecidableEq n] [Fintype n] (R : Type v) [CommRing R]
+
+/-- A projective special linear group is the quotient of a special linear group by its center.-/
+abbrev ProjectiveSpecialLinearGroup :=
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/ProjectiveSpecialLinearGroup.lean:23]:
```suggestion
abbrev ProjectiveSpecialLinearGroup : Type _ :=
```

**Comment by digama0** [comment]
Note: I have pushed an update to the lean toolchain because this PR was on a [buggy version](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Potential.20data.20loss.20from.20.60lake.20clean.60.20with.204.2E2.2E0-rc2.2F3/near/397875701) of the toolchain. **WARNING**: checking out old commits of this PR using v4.2.0-rc2 or v4.2.0-rc3 can cause `lake clean` to delete your mathlib folder! If you need to do so, make sure to delete `lakefile.olean` manually before running any `lake` commands.

**Comment by Ruben-VandeVelde** [comment]
This is a pretty bare definition, but unless someone has suggestions on what to add, I think we could merge it.

maintainer merge

File: Mathlib/LinearAlgebra/Matrix/ProjectiveSpecialLinearGroup.lean [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+open Matrix LinearMap
+
+open scoped MatrixGroups
+
+variable (n : Type u) [DecidableEq n] [Fintype n] (R : Type v) [CommRing R]
+
+/-- A projective special linear group is the quotient of a special linear group by its center.-/
+abbrev ProjectiveSpecialLinearGroup : Type _ :=
+    SpecialLinearGroup n R ⧸ Subgroup.center (SpecialLinearGroup n R)
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/ProjectiveSpecialLinearGroup.lean:30]: Is `X / Subgroup.Center X` something worth naming in its own right?

Comment (riccardobrasca) [reply Mathlib/LinearAlgebra/Matrix/ProjectiveSpecialLinearGroup.lean:30]: I think there is no standard name for this, so it would only confuse people.

**Comment by riccardobrasca** [comment]
bors merge

**Pull request closed without merging**